mockgen/model/model_gotypes.go

```diff
@@ -0,0 +1,161 @@
+package model
+
+import (
+	"fmt"
+	"go/types"
+)
+
+// InterfaceFromGoTypesType returns a pointer to an interface for the
+// given reflection interface type.
+func InterfaceFromGoTypesType(it *types.Interface) (*Interface, error) {
+	intf := &Interface{}
+
+	for i := 0; i < it.NumMethods(); i++ {
+		mt := it.Method(i)
+		// Skip unexported methods.
+		if !mt.Exported() {
+			continue
+		}
+		m := &Method{
+			Name: mt.Name(),
+		}
+
+		var err error
+		m.In, m.Variadic, m.Out, err = funcArgsFromGoTypesType(mt.Type().(*types.Signature))
+		if err != nil {
+			return nil, err
+		}
+
+		intf.AddMethod(m)
+	}
+
+	return intf, nil
+}
+
+func funcArgsFromGoTypesType(t *types.Signature) (in []*Parameter, variadic *Parameter, out []*Parameter, err error) {
+	nin := t.Params().Len()
+	if t.Variadic() {
+		nin--
+	}
+	var p *Parameter
+	for i := 0; i < nin; i++ {
+		p, err = parameterFromGoTypesType(t.Params().At(i), false)
+		if err != nil {
+			return
```

Comment:
optional nit:
Putting `var p` outside the loop feels unnecessary.
Hygiene suggests that the three blocks (for params, if, for results) don't have much shared.

We could scope each `p` to its own loop with `:=`.
(Looks like the naked `return`s are the only reason to do this which doesn't feel like a strong reason. The balance there is between the convenience of the naked return and the hygiene of isolating the three blocks, and I lean the latter, but I'll make it your call.)

```suggestion
	for i := 0; i < nin; i++ {
		p, err := parameterFromGoTypesType(t.Params().At(i), false)
		return nil, nil, err
```